test/unit/knex.js

```diff
@@ -98,10 +99,76 @@ describe('knex', () => {
     const knexWithParams = knex.withUserParams({ userParam: '451' });
 
     expect(knexWithParams.migrate.knex.userParams).to.deep.equal({
+      isProcessingDisabled: true,
+      postProcessResponse: undefined,
       userParam: '451',
+      wrapIdentifier: undefined,
     });
   });
 
+  it('copying does not result in duplicate listeners', () => {
+    if (isNode6()) {
+      return;
+    }
+
+    const knex = Knex({
+      client: 'sqlite',
+    });
+    const knexWithParams = knex.withUserParams();
+
+    expect(knex.client.listeners('start').length).to.equal(1);
+    expect(knex.client.listeners('query').length).to.equal(1);
+    expect(knex.client.listeners('query-error').length).to.equal(1);
+    expect(knex.client.listeners('query-response').length).to.equal(1);
+
+    expect(knexWithParams.client.listeners('start').length).to.equal(1);
+    expect(knexWithParams.client.listeners('query').length).to.equal(1);
+    expect(knexWithParams.client.listeners('query-error').length).to.equal(1);
+    expect(knexWithParams.client.listeners('query-response').length).to.equal(
+      1
+    );
+  });
+
+  it('listeners added to knex directly get copied correctly', () => {
+    const knex = Knex({
+      client: 'sqlite',
+    });
+    const onQueryResponse = function(response, obj, builder) {};
+    expect(knex.listeners('query-response').length).to.equal(0);
+    knex.on('query-response', onQueryResponse);
+
+    const knexWithParams = knex.withUserParams();
+
+    expect(knex.listeners('query-response').length).to.equal(1);
+    expect(knexWithParams.listeners('query-response').length).to.equal(1);
+  });
+
+  it('adding listener to copy does not affect base knex', () => {
+    if (isNode6()) {
+      return;
+    }
+
+    const knex = Knex({
+      client: 'sqlite',
+    });
+
+    expect(knex.client.listeners('start').length).to.equal(1);
+    expect(knex.client.listeners('query').length).to.equal(1);
+    expect(knex.client.listeners('query-error').length).to.equal(1);
+    expect(knex.client.listeners('query-response').length).to.equal(1);
+
+    const knexWithParams = knex.withUserParams();
+    knexWithParams.client.on('query', (obj) => {
```

Comment:
I'm starting to have second thoughts if this was a good idea to allow creating modifiable clones of knex instance which are sharing the same pool (feature came kind of accidentally). I think it is a really nice feature to have, but API could have separated cloning and setting same pools for them and setting user parameters of each instance more clearly... anyways test looks good 👍 Wanted to just mention how this small feature of user parameters has exploded a bit for allowing to set custom event handlers too, which is good.